### ISSUE TITLE
Network dispatch doesn't handle PoX invalidations

### ISSUE THREAD

**Issue by kantai** [stacks-network/stacks-blockchain#2701]
The `PeerNetwork::run` method can error in the event of PoX invalidations. This mostly only happens on testnet (where PoX cycles are _much_ more frequent, so nodes do occasionally perform a PoX "reorg"), but it's possible on mainnet too. Usually this isn't a problem, if `PeerNetwork::run` fails, the loop just keeps running. However, when the node is "event driven" (i.e., the event dispatcher is connected to an event consumer) like in the case of an API node, `PeerNetwork::run` errors are panicking. That's because the event dispatcher cannot recover from an arbitrary PeerNetwork error.

The code path that causes issues in testnet followers is `get_block_availability`:

```rust
        // what blocks do we have in this range?
        let local_blocks = {
            let ic = sortdb.index_conn();
            let tip = SortitionDB::get_canonical_burn_chain_tip(&ic)?;

...

            let local_blocks = ic.get_stacks_header_hashes(
                sortition_height_end - sortition_height_start,
                &last_ancestor.consensus_hash,
                header_cache,
            )?;
```

The bug witnessed is as follows:

1. `get_canonical_burn_chain_tip` is called and returns the then-canonical sortition tip.
2. the chains coordinator detects a PoX reorg and invalidates the sortition tip
3. `get_stacks_header_hashes` is called and returns an `Err` because the supplied tip is marked `pox_valid = 0`.

The invalidations are performed in a sqlite transaction, so they are atomic. However, 1 and 3 are not, these operations are performed on just an `index_conn`, not a sqlite transation. So perhaps the solution is to use a transaction here.

**Comment by jcnelson** [comment]
I see the problem, but I'd hesitate to put this into a transaction because we'd want to avoid database contention while bootstrapping (the sortition DB needs to have a transaction open when processing both a Bitcoin block and a Stacks block).  Maybe there's some optimistic concurrency control we could do here?

**Comment by jcnelson** [comment]
Path forward: implement something akin to an `EAGAIN` that the dispatcher knows to simply ignore.

**Comment by jcnelson** [comment]
Fixed in #2746
